### DISCUSSION TITLE
Regarding Webots and the python binding timing issue

### DISCUSSION THREAD

**Discussion by Maaaackster** [bitcraze/discussions#1044] [Q&A]
I'm trying to develop my own Crazyflie simulation in Webots based on the work already being done in this repo:

https://github.com/bitcraze/crazyflie-simulation

I've also integrated the simulation into ROS2. When trying to run the drone in setpoint mode with xyz absolute, I find that the drone behavior is very unstable for certain setpoints that I've verified work on the real drone. I now believe the timing issue, as mentioned here: https://github.com/bitcraze/crazyflie-firmware/issues/1320, to be the problem. After examining the stabilizer.c file from the drone firmware, I see the issue with the fixed ticks in the loop. I'm also aware that the fix for now for this issue is just to set the tick to a fixed 100, to always have the position and attitude PID running. But, as mentioned previously, this only seems to work fine for some cases. The natural solution, to my understanding, would be to implement a tick timer in the simulation loop and increment it for each iteration, passing that into the controller. While also lowering the time step in Webots to 1 ms. To my understanding, this should pretty well mimic what actually happens in the stabilizer loop in the actual firmware. Trying this, however, gives even worse results than using a fixed tick and larger timesteps in Webots. I simply don't understand why this would be the case. Any advice on this would be greatly appreciated.

**Comment by knmcguire** [comment]
Yeah... we tried to put it to 1 ms as well but it was still unstable. We haven't gotten the time to really dig into it.

I think 2 things are going on:
- There is more going on with the bindings than just the 'timing issue'
- The external controller of ROS is not giving the same outputs as expected

Unfortunately, that's how far we have gotten and I don't think that I'll have the time to look into it in more detail before the end of this year... If you don't mind which controller is used you can use the standard pid controller that is not based on the firmware, that is also an option. You can also contact webots to ask for some guidance of intergration. 

Or... take a look at this cool discussion of someone who achieved proper SITL with gazebo: https://github.com/orgs/bitcraze/discussions/995
